# Price analysis for 8 months (Nov22-June23)

In [1]:
import numpy as np
import pandas as pd

#!pip install pandas fastparquet

URL_DATA1 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2022-11.parquet'
URL_DATA2 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2022-12.parquet'
URL_DATA3 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2023-01.parquet'
URL_DATA4 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2023-02.parquet'
URL_DATA5 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2023-03.parquet'
URL_DATA6 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2023-04.parquet'
URL_DATA7 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2023-05.parquet'
URL_DATA8 = 'https://storage.googleapis.com/dosm-public-pricecatcher/pricecatcher_2023-06.parquet'

df1 = pd.read_parquet(URL_DATA1)
if 'date' in df1.columns: df1['date'] = pd.to_datetime(df1['date'])

df2 = pd.read_parquet(URL_DATA2)
if 'date' in df2.columns: df2['date'] = pd.to_datetime(df2['date'])

df3 = pd.read_parquet(URL_DATA3)
if 'date' in df3.columns: df3['date'] = pd.to_datetime(df3['date'])

df4 = pd.read_parquet(URL_DATA4)
if 'date' in df4.columns: df4['date'] = pd.to_datetime(df4['date'])

df5 = pd.read_parquet(URL_DATA5)
if 'date' in df5.columns: df5['date'] = pd.to_datetime(df5['date'])

df6 = pd.read_parquet(URL_DATA6)
if 'date' in df6.columns: df6['date'] = pd.to_datetime(df6['date'])

df7 = pd.read_parquet(URL_DATA7)
if 'date' in df7.columns: df7['date'] = pd.to_datetime(df7['date'])

df8 = pd.read_parquet(URL_DATA8)
if 'date' in df8.columns: df8['date'] = pd.to_datetime(df8['date'])


In [2]:
URL_DATA = 'https://storage.googleapis.com/dosm-public-pricecatcher/lookup_premise.parquet'
df_premise_code = pd.read_parquet(URL_DATA)
if 'date' in df_premise_code.columns: df_premise_code['date'] = pd.to_datetime(df_premise_code['date'])

print(f'There are {df_premise_code["district"].nunique()} different districts')
print(f'There are {df_premise_code["state"].nunique()} different states')
#print(df_premise_code.head(10))



URL_DATA = 'https://storage.googleapis.com/dosm-public-pricecatcher/lookup_item.parquet'
df_item_code = pd.read_parquet(URL_DATA)
if 'date' in df_item_code.columns: df_item_code['date'] = pd.to_datetime(df_item_code['date'])

print(f'There are {df_item_code["item"].nunique()} different items')
print(f'There are {df_item_code["item_category"].nunique()} different item_categories')
#print(df_item_code.head())


item_category_list = df_item_code["item_category"].unique()
item_list = df_item_code["item"].unique()

print(item_category_list)

print(item_list)


There are 173 different districts
There are 16 different states
There are 710 different items
There are 59 different item_categories
[None 'AYAM' 'DAGING' 'BUAH-BUAHAN' 'BAHAN LAUT' 'IKAN DARAT'
 'SAYUR-SAYURAN' 'KELAPA' 'TELUR' 'BAWANG' 'HASIL LAUT KERING' 'KACANG'
 'REMPAH RATUS (TIDAK BERBUNGKUS)' 'UBI KENTANG' 'BAHAN-BAHAN MINUMAN'
 'IKAN DALAM TIN' 'SAPUAN (SPREADS)' 'KICAP DAN SOS' 'TERSEDIA MINUM'
 'MINYAK DAN LEMAK' 'ROTI' 'PENJAGAAN DIRI' 'SUSU BAYI'
 'KRIMER DAN SUSU TEPUNG' 'REMPAH RATUS (BERBUNGKUS)' 'BERAS' 'TEPUNG'
 'UBAT-UBATAN' 'MOUTH WASH' 'MAKANAN SEGERA' 'COKLAT' 'ESEN DAN RAGI'
 'MAKANAN BAYI' 'SANTAN (KOTAK)' 'NASI' 'MEE / BIHUN / KUEY TEOW' 'LAUK'
 'LAIN-LAIN' 'MINUMAN' 'MEE/KUETIAU' 'BIHUN' 'LAMPIN PAKAI BUANG'
 'PENJAGAAN RUMAH' 'GULA' 'MENTEGA' 'ALAT TULIS DAN BAHAN BACAAN'
 'MAJALAH' 'BERUS GIGI' 'UBAT GIGI' 'TUALA WANITA' 'SYAMPU' 'TISU'
 'PENGHALAU NYAMUK' 'MAKANAN RINGAN' 'BISKUT' 'SABUN BADAN'
 'PEWANGI RUMAH' 'CILI KERING' 'MI SEGERA' 'TAUHU DAN TEMPE']
[

In [3]:

def map_item_premise_code(dataframe):
    tempdf = dataframe.copy() 
    
    s =tempdf.pop('premise_code').map(df_premise_code.set_index('premise_code')["state"])
    s.name = 'state'

    tempdf = dataframe.copy() 

    t = tempdf.pop('premise_code').map(df_premise_code.set_index('premise_code')['district'])
    t.name = "district"

    tempdf = dataframe.copy() 

    u = tempdf.pop('item_code').map(df_item_code.set_index('item_code')['item'])
    u.name = "item"

    tempdf = dataframe.copy() 

    v = tempdf.pop('item_code').map(df_item_code.set_index('item_code')['unit'])
    v.name = "unit"

    tempdf = dataframe.copy() 

    w = tempdf.pop('item_code').map(df_item_code.set_index('item_code')['item_category'])
    w.name = "item_category"

    df_master = pd.concat([dataframe,s,t,u,v,w],axis=1)
    df_master.sample(frac=1)

    return df_master

df1_master = map_item_premise_code(df1)
df2_master = map_item_premise_code(df2)
df3_master = map_item_premise_code(df3)
df4_master = map_item_premise_code(df4)
df5_master = map_item_premise_code(df5)
df6_master = map_item_premise_code(df6)
df7_master = map_item_premise_code(df7)
df8_master = map_item_premise_code(df8)


In [4]:
x = df7_master.query('state == "Johor" and item_category == "IKAN DALAM TIN"')
x.head()

date  premise_code  item_code  price  state     district  \
273264 2023-05-07           585        190    4.4  Johor      Pontian   
273529 2023-05-07           589        190    5.0  Johor      Pontian   
273530 2023-05-07           589        191    9.5  Johor      Pontian   
273740 2023-05-07           599        190    4.2  Johor  Kota Tinggi   
273741 2023-05-07           599        191    8.9  Johor  Kota Tinggi   

                                item   unit   item_category  
273264  SARDIN CAP AYAM (SOS TOMATO)  155 g  IKAN DALAM TIN  
273529  SARDIN CAP AYAM (SOS TOMATO)  155 g  IKAN DALAM TIN  
273530  SARDIN CAP AYAM (SOS TOMATO)  425 g  IKAN DALAM TIN  
273740  SARDIN CAP AYAM (SOS TOMATO)  155 g  IKAN DALAM TIN  
273741  SARDIN CAP AYAM (SOS TOMATO)  425 g  IKAN DALAM TIN

The code below confirms that the price for minyak masak tulen cap alif 5kg is as per myprice dashboard. The excel database 120 harga barang has a different
price for the mean price

In [5]:

l = [254, 257, 260, 263, 266, 1095, 1936, 1939, 1942]
l = [1942] #Minyak masak tulen cap alif 5kg
for item_codex in l:
    x = df7_master.query('state == "Johor" and item_code==@item_codex')
    #print(x.info())
    #x.hist(column='price')
    print(f'{x["price"].mean()}')
    #print(x["price"].mode())

#x = df_master.query('state=="Johor" and item_category=="IKAN DALAM TIN"')
#print(x["price"].mean())


30.24338150289017


In [6]:
'''
Officially-----
There are 173 different districts
There are 16 different states

There are 710 different items
There are 59 different item_categories


Results of June 2023-----

There are 163 different districts
There are 16 different states

There are 501 different items
There are 38 different item_categories
'''

print(f'There are {df7_master["item_code"].nunique()} different items')
print(f'There are {df7_master["item_category"].nunique()} different item_categories')

print(f'There are {df7_master["district"].nunique()} different districts')
print(f'There are {df7_master["state"].nunique()} different states')

item_list = df7_master["item"].unique()
district_list = df7_master["district"].unique()


def get_mean_price_monthly(dataframe_master):

    mean_price = []
    i=1
    for district in district_list:
        for item in item_list:
            newdf = dataframe_master.query('item==@item and district==@district')

            mean_price.append(newdf["price"].mean())

            #print(f'{i} Mean Price for {item} in {district} is {mean_price} ')

            i+=1
    
    return mean_price

#june23_mean_price = get_mean_price_monthly(df7_master)



There are 501 different items
There are 38 different item_categories
There are 163 different districts
There are 16 different states


Let's focus on Minyak Masak Tulen Cap Alif 5kg (code 1942)

Seems that it only has data for 106 districts. Which minyak masak 5kg would cover all 163 districts?



In [7]:
minyak5kg_list= [254, 257, 260, 263, 266, 1095, 1936, 1939, 1942]
    
def get_mean_price(dataframe_master,item_code):

    import math

    colLabel = ['District','Mean Price']
    mean_price = []

    i=1
    for district in district_list:
        newdf = dataframe_master.query('item_code==@item_code and district==@district')

        m_price = newdf["price"].mean()


        if math.isnan(m_price) == False: mean_price.append([district,m_price])
        
            
        i+=1
    
    pricedf = pd.DataFrame(mean_price,columns=colLabel)
    

    return pricedf

get_mean_price(df7_master,1942)

for item in minyak5kg_list: print(f' item_code:{item} mean_price:{len(get_mean_price(df4_master,item))}')    
for item in minyak5kg_list: print(f' item_code:{item} mean_price:{len(get_mean_price(df5_master,item))}')
for item in minyak5kg_list: print(f' item_code:{item} mean_price:{len(get_mean_price(df6_master,item))}')
for item in minyak5kg_list: print(f' item_code:{item} mean_price:{len(get_mean_price(df7_master,item))}')


 item_code:254 mean_price:146
 item_code:257 mean_price:119
 item_code:260 mean_price:149
 item_code:263 mean_price:147
 item_code:266 mean_price:148
 item_code:1095 mean_price:141
 item_code:1936 mean_price:68
 item_code:1939 mean_price:124
 item_code:1942 mean_price:101
 item_code:254 mean_price:145
 item_code:257 mean_price:122
 item_code:260 mean_price:150
 item_code:263 mean_price:151
 item_code:266 mean_price:147
 item_code:1095 mean_price:144
 item_code:1936 mean_price:70
 item_code:1939 mean_price:125
 item_code:1942 mean_price:102
 item_code:254 mean_price:147
 item_code:257 mean_price:126
 item_code:260 mean_price:149
 item_code:263 mean_price:151
 item_code:266 mean_price:151
 item_code:1095 mean_price:150
 item_code:1936 mean_price:71
 item_code:1939 mean_price:129
 item_code:1942 mean_price:106
 item_code:254 mean_price:150
 item_code:257 mean_price:129
 item_code:260 mean_price:149
 item_code:263 mean_price:151
 item_code:266 mean_price:153
 item_code:1095 mean_price:151


263 is Minyak Masak Tulen Cap Buruh 5kg. It seems that this particular minyak masak is recorded in 150 different districts.

What is the monthly trend of this minyak masak for each district?
Proof that the variation for this item is not so much from month to month. The variation is felt more district to district.

In [8]:
tempdf = df7_master.query('item_code==1942 and district=="Kinta"')

print(tempdf["price"].max())
print(tempdf["price"].min())

tempdf.head()


30.9
28.65


date  premise_code  item_code  price  state district  \
24270 2023-05-02          6039       1942  30.90  Perak    Kinta   
34464 2023-05-02         12352       1942  30.90  Perak    Kinta   
37296 2023-05-02         14807       1942  30.86  Perak    Kinta   
40130 2023-05-02         15903       1942  30.90  Perak    Kinta   
46930 2023-05-02         17636       1942  29.40  Perak    Kinta   

                              item  unit     item_category  
24270  MINYAK MASAK TULEN CAP ALIF  5 kg  MINYAK DAN LEMAK  
34464  MINYAK MASAK TULEN CAP ALIF  5 kg  MINYAK DAN LEMAK  
37296  MINYAK MASAK TULEN CAP ALIF  5 kg  MINYAK DAN LEMAK  
40130  MINYAK MASAK TULEN CAP ALIF  5 kg  MINYAK DAN LEMAK  
46930  MINYAK MASAK TULEN CAP ALIF  5 kg  MINYAK DAN LEMAK

In [9]:
def get_range_price(dataframe_master,item_code):

    import math

    colLabel = ['District','Range Price']
    range_price = []

    i=1
    for district in district_list:
        newdf = dataframe_master.query('item_code==@item_code and district==@district')

        range = newdf["price"].max()-newdf["price"].min()


        if math.isnan(range) == False: range_price.append([district,range])
        
            
        i+=1
    
    rangedf = pd.DataFrame(range_price,columns=colLabel)
    

    return rangedf

range_df = get_range_price(df7_master,263)

In [10]:
range_df[range_df["Range Price"]>0]

#df7_master.query('item_code==263 and district=="Kinta"')

District  Range Price
0            Kinta         2.91
1    Melaka Tengah         2.00
2           Kangar         1.91
3           Tampin         1.50
4           Cheras         2.91
..             ...          ...
132    Kuala Pilah         1.95
133    Kuala Nerus         2.91
134           Bera         2.91
139        Tangkak         2.91
145     Lubok Antu         1.00

[100 rows x 2 columns]